### PR TITLE
Add RMS norm operation across TTIR and TTNN dialects

### LINKED ISSUE #4262: Add ttnn.rms_norm op to TTNN IR
## Summary
Add support for `ttnn.rms_norm` operation to TTNN IR as part of LLM-specific ops onboarding for Llama3.2 architecture.

## Description
This op implements Root Mean Square (RMS) normalization, which is used instead of LayerNorm in many modern LLM architectures including Llama.

## Implementation Requirements
- [ ] Define the op in TTNN dialect
- [ ] Implement op lowering/conversion
- [ ] Add fusing patterns (investigation needed)
- [ ] Add unit tests
- [ ] Add integration tests
- [ ] Update documentation

## Related Issues
- Parent issue: #4214 (LLM-specific ops onboarding)

## Architecture
- Target: Llama3.2 architecture
- Category: Normalization operations

## Notes
- Fusing patterns will be needed for performance optimization
- Alternative to LayerNorm used in modern LLMs
- Critical normalization operation for Llama models
- Part of broader LLM ops initiative

### PR DESCRIPTION
**Pull request by sdjordjevicTT**
### Ticket
Closes #4262

### Problem description
The tt-mlir compiler was missing support for RMS (Root Mean Square) normalization operations, which are commonly used in modern neural network architectures like transformers. Without this operation, users could not compile models that utilize RMS normalization layers.

### What's changed
- Added TTIR RMSNormOp: New operation definition in TTIR dialect with support for normalized_shape, epsilon, optional weight/bias parameters, and comprehensive verification.
- Added TTNN RMSNormOp: Corresponding operation in TTNN dialect with memory config support and proper verification.
- Implemented conversion pipeline: Complete conversion from TTIR to TTNN with validation that normalized_shape matches TTNN's constraint of last-dimension normalization
- Added flatbuffer serialization: Full support for serializing RMS norm operations to flatbuffer format with proper handling of optional parameters
- Implemented runtime support: Complete runtime implementation including operation execution, tensor pool management, and proper handling of optional weight/bias tensors
- Added comprehensive tests: Test coverage across all levels - dialect verification, conversion pipeline, EmitC generation, and silicon execution tests

The implementation supports the full RMS normalization formula:
```
rms_norm(x, weight, bias, epsilon) = (x / sqrt(mean(x^2, dims=normalized_dims) + epsilon)) * weight + bias
```
with proper handling of optional weight and bias parameters.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 1]
```diff
@@ -4466,6 +4466,66 @@ ::mlir::LogicalResult mlir::tt::ttir::BatchNormOp::verify() {
...
+      return emitOpError("normalized_shape dimensions must match trailing "
+                         "dimensions of input tensor");
+    }
+  }
+
+  // Verify weight tensor shape if present.
+  if (getWeight()) {
+    const RankedTensorType weightType =
+        mlir::cast<RankedTensorType>(getWeight().getType());
+    if (weightType.getShape() != ArrayRef<int64_t>(normalizedShape)) {
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp]: Remove cast here and below.

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 2]
```diff
@@ -4466,6 +4466,66 @@ ::mlir::LogicalResult mlir::tt::ttir::BatchNormOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// RMSNormOp
+//===----------------------------------------------------------------------===//
+::mlir::LogicalResult mlir::tt::ttir::RMSNormOp::verify() {
+  const RankedTensorType inputType = getInput().getType();
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp]: The const is redundant for types/attrs/ops in mlir since they are value typed.

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 3]
```diff
@@ -4466,6 +4466,66 @@ ::mlir::LogicalResult mlir::tt::ttir::BatchNormOp::verify() {
...
+  for (size_t i = 0; i < normalizedShape.size(); ++i) {
+    if (inputShape[offset + i] != normalizedShape[i]) {
+      return emitOpError("normalized_shape dimensions must match trailing "
+                         "dimensions of input tensor");
+    }
+  }
+
+  // Verify weight tensor shape if present.
+  if (getWeight()) {
+    const RankedTensorType weightType =
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp]: Don't need cast

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 4]
```diff
@@ -2031,6 +2031,60 @@ ::mlir::LogicalResult mlir::tt::ttnn::BatchNormOp::verify() {
...
+::mlir::LogicalResult mlir::tt::ttnn::RMSNormOp::verify() {
+  RankedTensorType inputType = getInput().getType();
+  RankedTensorType outputType = getResult().getType();
+
+  // Input and output must have the same shape.
+  if (inputType.getShape() != outputType.getShape()) {
+    return emitOpError("input and output must have the same shape");
+  }
+
+  // For 0D tensors, weight and bias validation is different.
```
Comment (mtopalovicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:2151]: Is this needed? I'm not sure if properly support scalars in flatbuffer lowering and runtime.

File: test/python/golden/test_ttir_ops.py [hunk 5]
```diff
@@ -2534,3 +2534,122 @@ def dot_general_wrapper(
...
+    weight_tensor = None
+    bias_tensor = None
+
+    if weight is not None:
+        weight_tensor = torch.ones(normalized_shape, dtype=dtype)
+
+    if bias is not None:
+        bias_tensor = torch.zeros(normalized_shape, dtype=dtype)
+
+    # Manual RMS norm computation for golden reference
```
Comment (mtopalovicTT) [inline test/python/golden/test_ttir_ops.py]: Can we use torch implementation of [RMSNorm](https://docs.pytorch.org/docs/stable/generated/torch.nn.RMSNorm.html). The only diff is that you would need to add bias on it since torch version doesn't support bias.

File: test/python/golden/test_ttir_ops.py [hunk 6]
```diff
@@ -2534,3 +2534,122 @@ def dot_general_wrapper(
...
+            shape,
+            torch.float32,
+            normalized_shape,
+            weight=weight,
+            bias=bias,
+            epsilon=1e-5,
+            unit_attrs=unit_attrs,
+        )
+
+    # Skip cases where has_weight=True and has_bias=False
```
Comment (mtopalovicTT) [inline test/python/golden/test_ttir_ops.py]: Should we ping folks who own builder to support cases where multiple operands can be optional? Not sure how big of a change that could be...

**Review by mtopalovicTT — APPROVED** [review]

File: test/python/golden/test_ttir_ops.py (hunk 6 above)
Comment (tapspatel) [reply test/python/golden/test_ttir_ops.py]: do you have an example IR of what that would look like

Comment (mtopalovicTT) [reply test/python/golden/test_ttir_ops.py]: Here we have two optional tensors weight and bias:
https://github.com/tenstorrent/tt-mlir/pull/4301/files#diff-8f3930d72933276e68e5f1e48e833610d3b60fce54dbe1bfb9f4e6b1457c82b2R2102

If you take above definition than both of these are valid:

```mlir
ttir.rms(%input, %output, %weight, %bias)
```

```mlir
ttir.rms(%input, %output, %bias)
```

To map which input operand maps to which named operand we use `operandSegmentSizes`. So for the first example `operandSegmentSizes` would be `[1, 1, 1, 1]` but for second it would be `[1, 1, 0, 1]`.

File: test/python/golden/test_ttir_ops.py (hunk 5 above)
Comment (sdjordjevicTT) [reply test/python/golden/test_ttir_ops.py]: Agree, makes sense

File: lib/Dialect/TTNN/IR/TTNNOps.cpp (hunk 4 above)
Comment (sdjordjevicTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp:2151]: I obtained this information from the TT-Metal verifier, but I'm not entirely sure it makes sense. Regarding scalars, the runtime supports TTNN tensors with 0D. A month ago, I added support for reducing across all dimensions for reduction operations, so I believe this should work. Regardless, I will add a silicon test to confirm this.

File: test/python/golden/test_ttir_ops.py (hunk 6 above)
Comment (sdjordjevicTT) [reply test/python/golden/test_ttir_ops.py]: After some debugging and investigating, this works without any support; all MLIR build methods don't expect operandSegmentSizes as an attribute input, the operandSegmentSizes is built in the build method itself based on the provided inputs. In the ttir_builder infrastructure and Python part of the world, this work by supplying keyword arguments, in this specific example, weight and bias.

File: include/ttmlir/Target/TTNN/program.fbs [hunk 7]
```diff
@@ -76,6 +76,7 @@ union OpType {
   ReduceScatterOp,
   RepeatInterleaveOp,
   RepeatOp,
+  RMSNormOp,
```
Comment (jnie-TT) [inline include/ttmlir/Target/TTNN/program.fbs:81]: Nit: Maybe add RMSNormOp after reshape to maintain lexicographical order
